src/nspawn/nspawn.c

```diff
@@ -2796,6 +2827,10 @@ static int outer_child(
         if (r < 0)
                 return r;
 
+        r = setup_machine_id(directory);
```

Comment:
So... you ensure that we have a valid machine id here. `setup_machine_id` should sprout an output parameter, which should then be passed as input to `setup_journal`. This will avoid the ugliness of reading the file back from disk and parsing it again.